src/hotspot/share/runtime/vmThread.cpp

```diff
@@ -381,119 +287,198 @@ class HandshakeALotClosure : public HandshakeClosure {
+    tty->print_cr("VM self-destructed");
+    exit(-1);
+  }
+}
 
-    //
-    // Execute VM operation
-    //
-    { HandleMark hm(VMThread::vm_thread());
+void VMThread::inner_execute(VM_Operation* op) {
+  assert(Thread::current()->is_VM_thread(), "Must be the VM thread");
+
+  VM_Operation* prev_vm_operation = NULL;
+  if (_cur_vm_operation != NULL) {
+    // Check the VM operation allows nested VM operation.
+    // This normally not the case, e.g., the compiler
+    // does not allow nested scavenges or compiles.
+    if (!_cur_vm_operation->allow_nested_vm_operations()) {
+      fatal("Nested VM operation %s requested by operation %s",
+            op->name(), _cur_vm_operation->name());
+    }
+    op->set_calling_thread(_cur_vm_operation->calling_thread());
+    prev_vm_operation = _cur_vm_operation;
+  }
 
-      EventMark em("Executing VM operation: %s", vm_operation()->name());
-      assert(_cur_vm_operation != NULL, "we should have found an operation to execute");
+  _cur_vm_operation = op;
 
-      // If we are at a safepoint we will evaluate all the operations that
-      // follow that also require a safepoint
-      if (_cur_vm_operation->evaluate_at_safepoint()) {
-        log_debug(vmthread)("Evaluating safepoint VM operation: %s", _cur_vm_operation->name());
+  HandleMark hm(VMThread::vm_thread());
+  EventMark em("Executing %s VM operation: %s", prev_vm_operation != NULL ? "nested" : "", op->name());
 
-        SafepointSynchronize::begin();
+  log_debug(vmthread)("Evaluating %s %s VM operation: %s",
+                       prev_vm_operation != NULL ? "nested" : "",
+                      _cur_vm_operation->evaluate_at_safepoint() ? "safepoint" : "non-safepoint",
+                      _cur_vm_operation->name());
 
-        if (_timeout_task != NULL) {
-          _timeout_task->arm();
-        }
+  bool end_safepoint = false;
+  if (_cur_vm_operation->evaluate_at_safepoint() &&
+      !SafepointSynchronize::is_at_safepoint()) {
+    SafepointSynchronize::begin();
+    if (_timeout_task != NULL) {
+      _timeout_task->arm();
+    }
+    end_safepoint = true;
+  }
 
-        evaluate_operation(_cur_vm_operation);
-        _cur_vm_operation = NULL;
+  evaluate_operation(_cur_vm_operation);
 
-        if (_timeout_task != NULL) {
-          _timeout_task->disarm();
-        }
+  if (end_safepoint) {
+    if (_timeout_task != NULL) {
+      _timeout_task->disarm();
+    }
+    SafepointSynchronize::end();
+  }
 
-        // Complete safepoint synchronization
-        SafepointSynchronize::end();
+  _cur_vm_operation = prev_vm_operation;
+}
+
+void VMThread::wait_for_operation() {
+  assert(Thread::current()->is_VM_thread(), "Must be the VM thread");
+  MonitorLocker ml_op_lock(VMOperation_lock, Mutex::_no_safepoint_check_flag);
+
+  // Clear previous operation.
+  // On first call this clears a dummy place-holder.
+  _next_vm_operation = NULL;
+  // Notify operation done and notify a next operation can be installed.
```

Comment:
typo - s/operation done/operation is done/

Reply:
Fixed